### PR TITLE
Fix missing frees and double frees

### PR DESCRIPTION
**Pull request by jwerle**
This pull request also cleans up some code and ensures a strict contract between the caller and function. All input memory is owned by the caller and all output memory is owned by the function. Output memory provided to a callback is free'd after the callback frame

### CODE REVIEW

**Comment by mribbons** [comment]
Runs on windows 👍

**Review by mribbons — APPROVED** [review]

**Review by mribbons — CHANGES_REQUESTED** [review]

File: bin/install.sh [hunk 1]
```diff
@@ -429,7 +429,7 @@ function _check_compiler_features {
     cflags+=(-o /dev/null)
   fi
 
-  $CXX -x c++ "${cflags[@]}" - >/dev/null << EOF_CC
+  $CXX -x c++ "${cflags[@]}" - -o /dev/null >/dev/null << EOF_CC
```
[the remark this answers is not in the document]
Comment (jwerle) [reply bin/install.sh:432]: I think we'd want to see this output upon failure, even without `VERBOSE` - it'll give a pretty obvious reason why the build failed during this step

File: test/src/test-context.js [hunk 2]
```diff
@@ -2,11 +2,12 @@ import { GLOBAL_TEST_RUNNER } from 'socket:test'
...
+import ipc from 'socket:ipc'
 
-// uncomment below to get IPC debug output in stdout
-// import ipc from 'socket:ipc'
-// ipc.debug.enabled = true
-// ipc.debug.log = (...args) => console.log(...args)
+if (process.env.DEBUG || process.env.DEBUG_IPC) {
+  ipc.debug.enabled = true
+  ipc.debug.log = (...args) => console.log(...args)
+}
```
[the remark this answers is not in the document]
Comment (jwerle) [reply test/src/test-context.js:10]: This works for now - maybe we support `ssc build --debug` one day and it sets the `DEBUG` env var

**Review by chicoxyzzy — APPROVED** [review]

File: bin/install.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mribbons) [reply bin/install.sh:432]: Fair point. I made a change in #166 so quiet outputs the command with `VERBOSE`, 
This one needs the command output as well, especially on failure.

File: test/src/test-context.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mribbons) [reply test/src/test-context.js:10]: Can we limit it to just DEBUG_IPC? Otherwise this is still going to flood the output...

**Review by heapwolf — APPROVED** [review]

File: test/src/test-context.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mribbons) [reply test/src/test-context.js:10]: Can we limit it to just DEBUG_IPC? Otherwise this is still going to flood the output...

**Review by mribbons — CHANGES_REQUESTED** [review]

**Review by trevnorris — CHANGES_REQUESTED** [review]
Amazing job. Thanks for being so thorough. Just need a couple of tweeks.

**Comment by chicoxyzzy** [comment]
can you also rebase on master to make CI work?

File: src/ipc/bridge.cc [hunk 3]
```diff
@@ -1151,20 +1168,48 @@ static void registerSchemeHandler (Router *router) {
...
+
       webkit_uri_scheme_request_finish_with_response(request, response);
+      g_input_stream_close_async(stream, 0, nullptr, +[](
+        GObject* object,
+        GAsyncResult* res,
+        gpointer userData
+      ) {
+        g_object_ref(res);
+        auto stream = (GInputStream*) object;
+        auto data = (const char*) userData;
```
[the remark this answers is not in the document]
Comment (jwerle) [reply src/ipc/bridge.cc:1200]: lets do a sweep of this later. I loathe c++ style casts, but it makes sense

**Pull request merged**